### PR TITLE
Compile phase inside chroot

### PR DESCRIPTION
**Pull request by thijskh**
This is an approach to put the compile phase inside the chroot. It works for me with C/C++ and Java (detect).

By and large this now works. I think actually it does not make things much more complex and I've managed to clean some things up a little even.

Some further improvements are thinkable, it could be e.g. optimised a little more to clean up the compile scripts from the chroot and symlink them back, as we do in testcase run.

### CODE REVIEW

File: judge/runguard.c [hunk 1]
```diff
@@ -828,8 +831,7 @@ void setrestrictions()
 		free(path);
 
 		if ( chroot(".")!=0 ) error(errno,"cannot change root to `%s'",cwd);
-		/* Just to make sure and satisfy Coverity scan: */
-		if ( chdir("/")!=0 ) error(errno,"cannot chdir to `/' in chroot");
+		if ( chdir(rootchdir)!=0 ) error(errno,"cannot chdir to `%s' in chroot", rootchdir);
```
Comment (eldering) [inline judge/runguard.c:834]: We should check here that `rootchdir` is an absolute path. Part of the reason for doing the `chdir` is to make sure that we're now inside the chroot. I'm not sure what the effect would be if we were not inside the chroot and follow a relative path...

Comment (thijskh) [reply judge/runguard.c:834]: Or just prepend “/“ to its value in this call of chdir

File: judge/runguard.c [hunk 2]
```diff
@@ -349,6 +351,7 @@ Run COMMAND with restrictions.\n\
   -r, --root=ROOT        run COMMAND with root directory set to ROOT\n\
   -u, --user=USER        run COMMAND as user with username or ID USER\n\
   -g, --group=GROUP      run COMMAND under group with name or ID GROUP\n\
+  -d, --chdir=DIR        change to directory DIR after setting root directory\n\
```
Comment (meisterT) [inline judge/runguard.c:354]: why can't we switch directory before calling runguard?

File: sql/files/defaultdata/java_javac_detect/run [hunk 3]
```diff
@@ -56,9 +56,9 @@ rm -f $TMPFILE
 
 if [ -z "$MAINCLASS" ]; then
 	# Look for class that has the 'main' function:
-	CLASSNAMES="$(find ./* -type f -regex '^.*\.class$' \
+	CLASSNAMES="$(find ./* -type f -regex '^.*\.class$' ! -name DetectMain.class \
```
Comment (meisterT) [inline sql/files/defaultdata/java_javac_detect/run:59]: this looks like something we should do regardless of this change, right?

File: judge/runguard.c (hunk 1 above)
Comment (thijskh) [reply judge/runguard.c:834]: Or keep the original chdir(/) and then call the chdir(rootchdir) after it

File: judge/runguard.c (hunk 2 above)
Comment (thijskh) [reply judge/runguard.c:354]: Because runguard does the chroot()ing. So if we chdir before we’re not in the dir with the right bind mounts.

Or put differently, the chroot is started once for a judging, which means both the compile and the run phase. So we use a different subdir for compile to not make it too messy

File: sql/files/defaultdata/java_javac_detect/run (hunk 3 above)
Comment (thijskh) [reply sql/files/defaultdata/java_javac_detect/run:59]: No it’s only now a problem because we copy this class in, before we included it from a different part of the fs but of course this is no longer readable when in the chroot

File: doc/manual/install-judgehost.rst [hunk 4]
```diff
@@ -76,23 +72,24 @@ paths, be sure to update the sudoers rules accordingly.
 Creating a chroot environment
 -----------------------------
 
-The judgedaemon executes submissions inside a chroot environment for
-security reasons. By default it mounts parts of a prebuilt chroot tree
-read-only during this judging process (using the script
-``lib/judge/chroot-startstop.sh``). This is needed to support
-extra languages that require access to interpreters or support
+The judgedaemon compiles and executes submissions inside a chroot
+environment for securwty reasons. By default it mounts parts of a
```
Comment (eldering) [inline doc/manual/install-judgehost.rst]: Typo `securwty`.

File: misc-tools/dj_make_chroot.in [hunk 5]
```diff
@@ -133,15 +133,15 @@ if [ "$DISTRO" = 'Debian' ]; then
 INCLUDEDEBS="ca-certificates"
 
 # Packages to install after upgrade (space separated):
-INSTALLDEBS="default-jre-headless pypy python3 locales"
+INSTALLDEBS="gcc g++ default-jdk default-jre-headless pypy python3 locales"
```
Comment (eldering) [inline misc-tools/dj_make_chroot.in:136]: Shouldn't we install `default-jdk-headless`?

File: doc/manual/install-judgehost.rst [hunk 6]
```diff
@@ -76,23 +72,24 @@ paths, be sure to update the sudoers rules accordingly.
...
+prebuilt chroot tree read-only during this judging process (using
+the script ``lib/judge/chroot-startstop.sh``). This is needed to
+support languages that require access to interpreters or support
```
Comment (eldering) [inline doc/manual/install-judgehost.rst:79]: This should be rewritten as it still seems to imply that a chroot environment is only needed for interpreted languages, while now it's going to be needed for any language, compiled or interpreted (or anything in between, like Java).

File: judge/build_executable.sh [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	$DJ_LIBJUDGEDIR/chroot-startstop.sh stop
+
+	# Make sure that all files are owned by the current user/group, so
+	# that we can delete the judging output tree without root access.
+	# We also remove group RUNGROUP so that this can safely be shared
+	# across multiple judgedaemons, and remove write permissions.
+	$GAINROOT chown -R "$(id -un):" $WORKDIR
+	chmod -R go-w $WORKDIR
+	for i in bin etc lib lib64 proc usr; do
+		rm -rf $i
```
Comment (eldering) [inline judge/build_executable.sh:23]: Why is this necessary? It happened to me that `proc` was not unmounted (probably because of an error in a previous run), and then I got an endless list of permission denied errors (and maybe some files deleted).

Comment (thijskh) [reply judge/build_executable.sh:23]: Because otherwise the (empty) dirs are also copied into each judging dir which clutters things.

Comment (eldering) [reply judge/build_executable.sh:23]: Can't we build/compile in a subdirectory and only copy contents of that directory? (I haven't looked into all the details yet of your PR.)

Comment (thijskh) [reply judge/build_executable.sh:23]: This is indeed possible I think
